graylog2-server/src/main/java/org/graylog2/indexer/messages/Messages.java

```diff
@@ -137,40 +142,88 @@ public ResultMessage get(String messageId, String index) throws DocumentNotFound
             return Collections.emptyList();
         }
 
-        final Bulk.Builder bulk = new Bulk.Builder();
-        for (Map.Entry<IndexSet, Message> entry : messageList) {
-            final Message message = entry.getValue();
-            if (isSystemTraffic) {
-                systemTrafficCounter.inc(message.getSize());
-            } else {
-                outputByteCounter.inc(message.getSize());
+        int chunkSize = messageList.size();
+        List<BulkResult.BulkResultItem> failedItems = new ArrayList<>();
+        for (;;) {
+            try {
+                failedItems = bulkIndexChunked(messageList, isSystemTraffic, chunkSize);
+                break; // on success
+            } catch (EntityTooLargeException e) {
+                LOG.warn("Bulk index failed with 'Request Entity Too Large' error. Retrying by splitting up batch size <{}>.", chunkSize);
+                if (chunkSize == messageList.size()) {
+                    LOG.warn("Consider lowering the \"output_batch_size\" setting.");
+                }
+                chunkSize /= 2;
+            }
+            if (chunkSize == 0) {
+                throw new ElasticsearchException("Bulk index cannot split output batch any further.");
             }
-
-            bulk.addAction(new Index.Builder(message.toElasticSearchObject(invalidTimestampMeter))
-                .index(entry.getKey().getWriteIndexAlias())
-                .type(IndexMapping.TYPE_MESSAGE)
-                .id(message.getId())
-                .build());
         }
 
-        final BulkResult result = runBulkRequest(bulk.build(), messageList.size());
-        final List<BulkResult.BulkResultItem> failedItems = result.getFailedItems();
-
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("Index: Bulk indexed {} messages, took {} ms, failures: {}",
-                    result.getItems().size(), result, failedItems.size());
-        }
 
         if (!failedItems.isEmpty()) {
             final Set<String> failedIds = failedItems.stream().map(item -> item.id).collect(Collectors.toSet());
             recordTimestamp(messageList, failedIds);
-            return propagateFailure(failedItems, messageList, result.getErrorMessage());
+            return propagateFailure(failedItems, messageList);
         } else {
             recordTimestamp(messageList, Collections.emptySet());
             return Collections.emptyList();
         }
     }
 
+    private List<BulkResult.BulkResultItem> bulkIndexChunked(final List<Map.Entry<IndexSet, Message>> messageList, boolean isSystemTraffic, int chunkSize) throws EntityTooLargeException {
+        chunkSize = Math.min(messageList.size(), chunkSize);
+
+        final List<BulkResult.BulkResultItem> failedItems = new ArrayList<>();
+        final Iterable<List<Map.Entry<IndexSet, Message>>> partition = Iterables.partition(messageList, chunkSize);
+        int partitionCount = 1;
+        for (List<Map.Entry<IndexSet, Message>> subMessageList: partition) {
```

Comment:
I think this implementation assumes that the messages inside `messageList` are uniformly sized.

Let's say we have a body limit of 10 MB and we have a `messageList` with message sizes of `[1 MB, 1 MB, 1 MB, 10 MB]`. If we run this, we will get a `413` for the first try and split the message list into two partitions. The first partition (two messages) will be successfully written to ES but for the second partition we will get a `413` response from ES again. Then we throw an `EntityTooLargeException` and start over indexing **all** messages again.

This will result in duplicated messages.

One possible solution could be to store the offset of the successfully written messages in the `EntityTooLargeException` and then use `List#subList()` to only pass in the remaining messages into `bulkIndexChunked()`.
This would still not handle the issue that `failedItems` of successful bulk requests would be lost. (we could include them in the exception as well...)

There might be a better way.

Reply:
Good catch! I've made some changes to handle this.